File: app/src/main/java/com/sagalogistics/backend/models/OrderImpl.java
package com.sagalogistics.backend.models;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.firebase.database.Exclude;

import java.util.HashMap;
import java.util.Map;

public class OrderImpl implements Order, Parcelable{

    public static final Parcelable.Creator CREATOR = new Creator() {
        @Override
        public OrderImpl createFromParcel(Parcel in) {
            return new OrderImpl(in);
        }

        @Override
        public OrderImpl[] newArray(int size) {
            return new OrderImpl[size];
        }
    };

    private String key;
    private Map<String, Integer> items = new HashMap<>();

    public OrderImpl(){
    }

    public OrderImpl(String key, Map<String, Integer> items) {
        this.key = key;
        this.items = items;
    }

    public OrderImpl(Map<String, Integer> items){
        this.items = items;
    }

    @Exclude
    public String getKey() {
        return key;
    }

    @Exclude
    public void setKey(String key) {
        this.key = key;
    }

    public Map<String, Integer> getItems() {
        return new HashMap<>(items); //Creem una shallow copy per impedir modificacions a la ordre real
    }

    public void setItems(Map<String, Integer> items) {
        this.items = items;
    }

    public void updateItem(String key, Integer quantity) {
        items.put(key, quantity);
    }

    public Integer getQuantityItem(String key) {
        return items.get(key);
    }

    public void removeItem(String key) {
        items.remove(key);
    }

    protected OrderImpl(Parcel in) {
        this.key = in.readString();
        in.readMap(this.items, Map.class.getClassLoader());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.key);
        dest.writeMap(this.items);
    }
}

File: app/src/main/java/com/sagalogistics/backend/models/OrderImpl.kt
package com.sagalogistics.backend.models

import android.os.Parcelable
import com.google.firebase.database.Exclude
import android.os.Parcel
import android.os.Parcelable.Creator

class OrderImpl : Order, Parcelable {
    companion object CREATOR : Creator<OrderImpl> {
        override fun createFromParcel(parcel: Parcel): OrderImpl {
            return OrderImpl(parcel)
        }

        override fun newArray(size: Int): Array<OrderImpl?> {
            return arrayOfNulls(size)
        }
    }

    @get:Exclude
    @set:Exclude
    override var key: String? = null
    override var items: MutableMap<String, Int>
        get() = LinkedHashMap(_items)
        set(value) {
            _items = value
        }
    private var _items: MutableMap<String, Int> = LinkedHashMap()

    constructor() {}

    constructor(key: String?, items: MutableMap<String, Int>) {
        this.key = key
        this._items = items
    }

    constructor(items: MutableMap<String, Int>) {
        this._items = items
    }

    override fun updateItem(key: String, quantity: Int) {
        _items[key] = quantity
    }

    override fun getQuantityItem(key: String): Int? {
        return _items[key]
    }

    override fun removeItem(key: String) {
        _items.remove(key)
    }

    constructor(parcel: Parcel) {
        key = parcel.readString()
        parcel.readMap(_items, MutableMap::class.java.classLoader)
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(key)
        dest.writeMap(items)
    }
}